src/Microsoft.AspNetCore.Dispatcher/DispatcherBase.cs

```diff
@@ -117,6 +121,16 @@ protected virtual Task<bool> TryMatchAsync(HttpContext httpContext)
                 throw new ArgumentNullException(nameof(selectors));
             }
 
+            if (!_selectorsInitialized)
+            {
+                foreach (var selector in Selectors)
+                {
+                    selector.Initialize(this);
+                }
+
+                _selectorsInitialized = true;
+            }
```

Comment:
I think it would be better here to use LazyInitializer here: https://msdn.microsoft.com/en-us/library/dd642260(v=vs.110).aspx

You can just have it return null.

The problem with this pattern is that it can run multiple times when multiple threads come in before initialization completes. We've seen this in routing on ARM and I don't want it to start happening in the new codebase